File: app/lib/models/service_due.dart
import 'package:flutter/foundation.dart';

import 'site.dart';
import 'site_config.dart';

/// How urgent a service is.
enum DueStatus {
  /// Nothing to do yet.
  ok('OK'),

  /// Inside the site's reminder lead, by distance or by date.
  dueSoon('DUE SOON'),

  /// The interval has been passed.
  overdue('OVERDUE'),

  /// No usable odometer, so distance-driven plans cannot be judged.
  unknown('NO ODOMETER');

  const DueStatus(this.label);

  final String label;

  bool get needsAttention =>
      this == DueStatus.dueSoon || this == DueStatus.overdue;
}

/// When one vehicle's one service plan next falls due.
///
/// A service is due on whichever comes first, distance or elapsed time — the
/// way paid car servicing works — so both are computed and the worse one wins.
@immutable
class ServiceDue {
  const ServiceDue({
    required this.vehicle,
    required this.plan,
    required this.status,
    this.dueAtKm,
    this.kmRemaining,
    this.dueOn,
    this.daysRemaining,
  });

  final Vehicle vehicle;
  final ServicePlan plan;
  final DueStatus status;

  /// Odometer reading the service falls due at. Null for a time-only plan.
  final int? dueAtKm;

  /// Distance still to run. Negative means overdue by that much.
  final int? kmRemaining;

  /// `yyyy-MM-dd` the service falls due. Null for a distance-only plan.
  final String? dueOn;

  /// Days left. Negative means overdue by that many.
  final int? daysRemaining;

  bool get needsAttention => status.needsAttention;

  /// The shorter of the two runways, for sorting a work queue.
  ///
  /// Distance is converted at a nominal 200 km/day so km and days compare on
  /// one axis; it only has to order the list, not predict a date.
  int get urgencyScore {
    final byKm = kmRemaining == null ? null : (kmRemaining! ~/ 200);
    final byDays = daysRemaining;
    if (byKm == null) return byDays ?? 1 << 20;
    if (byDays == null) return byKm;
    return byKm < byDays ? byKm : byDays;
  }

  String get summary {
    final parts = <String>[
      if (kmRemaining != null)
        kmRemaining! >= 0
            ? 'in ${_thousands(kmRemaining!)} km'
            : '${_thousands(-kmRemaining!)} km over',
      if (daysRemaining != null)
        daysRemaining! >= 0
            ? 'in $daysRemaining days'
            : '${-daysRemaining!} days over',
    ];
    if (parts.isEmpty) return 'Not scheduled';
    return parts.join(' · ');
  }
}

/// Computes the maintenance queue for a site.
abstract final class ServiceSchedule {
  /// Every active plan against every active vehicle.
  ///
  /// A vehicle with no odometer reading yields [DueStatus.unknown] for
  /// distance-driven plans rather than being silently treated as at zero km —
  /// a stale telematics feed must be visible, not invisible.
  static List<ServiceDue> forSite({
    required List<Vehicle> vehicles,
    required SiteConfig config,
    required DateTime now,
  }) {
    final out = <ServiceDue>[];
    for (final vehicle in vehicles.where((v) => v.isActive)) {
      for (final plan in config.activePlans) {
        out.add(_evaluate(vehicle: vehicle, plan: plan, config: config, now: now));
      }
    }
    out.sort((a, b) => a.urgencyScore.compareTo(b.urgencyScore));
    return out;
  }

  /// Only the rows a manager has to act on.
  static List<ServiceDue> attentionOnly({
    required List<Vehicle> vehicles,
    required SiteConfig config,
    required DateTime now,
  }) =>
      forSite(vehicles: vehicles, config: config, now: now)
          .where((d) => d.needsAttention)
          .toList();

  static ServiceDue _evaluate({
    required Vehicle vehicle,
    required ServicePlan plan,
    required SiteConfig config,
    required DateTime now,
  }) {
    int? dueAtKm;
    int? kmRemaining;
    String? dueOn;
    int? daysRemaining;

    var status = DueStatus.ok;
    var distanceUsable = true;

    if (plan.isDistanceDriven) {
      if (!vehicle.hasOdometer) {
        distanceUsable = false;
      } else {
        // Anchor on the last service if there is one, otherwise on zero — a
        // vehicle that has never been serviced is due at its first interval.
        final anchor = vehicle.lastServiceKm ?? 0;
        dueAtKm = anchor + plan.intervalKm;
        kmRemaining = dueAtKm - vehicle.odometerKm;

        if (kmRemaining <= 0) {
          status = DueStatus.overdue;
        } else if (kmRemaining <= config.reminderLeadKm) {
          status = DueStatus.dueSoon;
        }
      }
    }

    if (plan.isTimeDriven) {
      // Anchor on the last service date, else the odometer's first sighting,
      // else today — the point is to start a clock, not to invent history.
      final anchor = _parseDate(vehicle.lastServiceOn) ??
          _parseDate(vehicle.odometerUpdatedAt) ??
          now;
      final due = anchor.add(Duration(days: plan.intervalDays));
      dueOn = _iso(due);
      daysRemaining = _wholeDaysBetween(now, due);

      final timeStatus = daysRemaining <= 0
          ? DueStatus.overdue
          : (daysRemaining <= config.reminderLeadDays
              ? DueStatus.dueSoon
              : DueStatus.ok);
      status = _worse(status, timeStatus);
    }

    // Only report "no odometer" when nothing else already demands attention.
    if (!distanceUsable && !status.needsAttention) {
      status = DueStatus.unknown;
    }

    return ServiceDue(
      vehicle: vehicle,
      plan: plan,
      status: status,
      dueAtKm: dueAtKm,
      kmRemaining: kmRemaining,
      dueOn: dueOn,
      daysRemaining: daysRemaining,
    );
  }

  static DueStatus _worse(DueStatus a, DueStatus b) {
    const rank = <DueStatus, int>{
      DueStatus.ok: 0,
      DueStatus.unknown: 1,
      DueStatus.dueSoon: 2,
      DueStatus.overdue: 3,
    };
    return rank[a]! >= rank[b]! ? a : b;
  }

  static DateTime? _parseDate(String? value) {
    if (value == null || value.isEmpty) return null;
    return DateTime.tryParse(value);
  }

  /// Calendar days between two instants, floored, ignoring the time of day so
  /// "due today" reads as 0 rather than a fraction.
  static int _wholeDaysBetween(DateTime from, DateTime to) {
    final a = DateTime(from.year, from.month, from.day);
    final b = DateTime(to.year, to.month, to.day);
    return b.difference(a).inDays;
  }

  static String _iso(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-'
      '${d.month.toString().padLeft(2, '0')}-'
      '${d.day.toString().padLeft(2, '0')}';
}

String _thousands(int value) {
  final s = value.abs().toString();
  final buffer = StringBuffer(value < 0 ? '-' : '');
  for (var i = 0; i < s.length; i++) {
    if (i > 0 && (s.length - i) % 3 == 0) buffer.write(',');
    buffer.write(s[i]);
  }
  return buffer.toString();
}

File: backend/app/services/service_due.py
from __future__ import annotations

from datetime import date as date_t
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.master import Vehicle
from app.models.site_config import ServicePlan, SiteConfig
from app.schemas.site import ServiceDueOut
from app.services.site_config import get_or_create, load_plans

#: Nominal km/day used only to put distance and time runways on one axis for
#: sorting. It orders the queue; it does not predict a date.
KM_PER_DAY = 200

_RANK = {"ok": 0, "unknown": 1, "due_soon": 2, "overdue": 3}


def _worse(a: str, b: str) -> str:
    return a if _RANK[a] >= _RANK[b] else b


def evaluate(
    vehicle: Vehicle, plan: ServicePlan, config: SiteConfig, today: date_t
) -> ServiceDueOut:
    """Mirrors `app/lib/models/service_due.dart` exactly — change both together."""
    due_km: int | None = None
    km_remaining: int | None = None
    due_on: date_t | None = None
    days_remaining: int | None = None

    status = "ok"
    distance_usable = True
    has_odometer = vehicle.odometer_updated_at is not None

    if plan.interval_km > 0:
        if not has_odometer:
            # A missing reading is unknown, never 0 km. A stale telematics feed
            # has to be visible, not invisible.
            distance_usable = False
        else:
            anchor = vehicle.last_service_km or 0
            due_km = anchor + plan.interval_km
            km_remaining = due_km - vehicle.odometer_km
            if km_remaining <= 0:
                status = "overdue"
            elif km_remaining <= config.reminder_lead_km:
                status = "due_soon"

    if plan.interval_days > 0:
        anchor_date = vehicle.last_service_on or _date_of(
            vehicle.odometer_updated_at, today
        )
        due_on = _add_days(anchor_date, plan.interval_days)
        days_remaining = (due_on - today).days
        if days_remaining <= 0:
            time_status = "overdue"
        elif days_remaining <= config.reminder_lead_days:
            time_status = "due_soon"
        else:
            time_status = "ok"
        status = _worse(status, time_status)

    # Only report "no odometer" when nothing else already demands attention.
    if not distance_usable and status not in ("due_soon", "overdue"):
        status = "unknown"

    return ServiceDueOut(
        vehicle_id=vehicle.id,
        registration_no=vehicle.registration_no,
        plan_code=plan.code,
        plan_name=plan.name,
        status=status,
        due_km=due_km,
        km_remaining=km_remaining,
        due_on=due_on,
        days_remaining=days_remaining,
        odometer_km=vehicle.odometer_km if has_odometer else None,
        has_odometer=has_odometer,
    )


def _date_of(value: datetime | None, fallback: date_t) -> date_t:
    return value.date() if value is not None else fallback


def _add_days(anchor: date_t, days: int) -> date_t:
    from datetime import timedelta

    return anchor + timedelta(days=days)


def _urgency(row: ServiceDueOut) -> int:
    by_km = None if row.km_remaining is None else row.km_remaining // KM_PER_DAY
    by_days = row.days_remaining
    if by_km is None:
        return by_days if by_days is not None else 1 << 20
    if by_days is None:
        return by_km
    return min(by_km, by_days)


async def for_site(
    session: AsyncSession, site_code: str, today: date_t
) -> list[ServiceDueOut]:
    """Every active plan against every active vehicle, worst first."""
    config = await get_or_create(session, site_code)
    plans = [
        p
        for p in await load_plans(session, site_code)
        if p.is_active and (p.interval_km > 0 or p.interval_days > 0)
    ]
    if not plans:
        return []

    vehicles = list(
        (
            await session.scalars(
                select(Vehicle)
                .where(Vehicle.site_code == site_code, Vehicle.is_active.is_(True))
                .order_by(Vehicle.registration_no)
            )
        ).all()
    )

    rows = [
        evaluate(vehicle, plan, config, today)
        for vehicle in vehicles
        for plan in plans
    ]
    rows.sort(key=_urgency)
    return rows
